sparse/impl/KokkosSparse_sort_crs_impl.hpp

```diff
@@ -0,0 +1,305 @@
+        // write out and reset
+        mergedEntries(insertPos) = accumCol;
+        insertPos++;
+        accumCol = entries(j);
+      }
+    }
+    // always left with the last unique entry
+    mergedEntries(insertPos) = accumCol;
+  }
+
+  typename rowmap_t::const_type rowmap;
+  entries_t entries;
+  rowmap_t mergedRowmap;
+  entries_t mergedEntries;
+};
+
+template <typename Offset, typename Keys, typename Entries>
+struct MaxScanFunctor {
+  using value_type = uint64_t;
+
+  MaxScanFunctor(uint64_t ncols_, const Keys& keys_, const Entries& entries_)
+      : ncols(ncols_), keys(keys_), entries(entries_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void init(uint64_t& update) const { update = 0; }
+
+  KOKKOS_INLINE_FUNCTION
+  void join(uint64_t& update, const uint64_t& input) const { update = Kokkos::max(update, input); }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(Offset i, uint64_t& lmax, bool finalPass) const {
+    lmax = Kokkos::max(lmax, keys(i));
+    if (finalPass) {
+      // lmax is the row containing entry i.
+      // The key is equivalent to the entry's linear
+      // index if the matrix were dense and row-major.
+      keys(i) = lmax * ncols + entries(i);
+    }
+  }
+
+  uint64_t ncols;
+  Keys keys;
+  Entries entries;
+};
+
+template <typename ExecSpace, typename Rowmap, typename Entries>
+Kokkos::View<uint64_t*, ExecSpace> generateBulkCrsKeys(const ExecSpace& exec, const Rowmap& rowmap,
+                                                       const Entries& entries,
+                                                       typename Entries::non_const_value_type ncols) {
+  using Offset    = typename Rowmap::non_const_value_type;
+  using Ordinal   = typename Entries::non_const_value_type;
+  Ordinal numRows = rowmap.extent(0) ? rowmap.extent(0) - 1 : 0;
+  Kokkos::View<uint64_t*, ExecSpace> keys("keys", entries.extent(0));
+  Kokkos::parallel_for(
+      "CRS bulk sorting: mark row begins", Kokkos::RangePolicy<ExecSpace>(exec, 0, numRows), KOKKOS_LAMBDA(Ordinal i) {
+        Offset rowBegin = rowmap(i);
+        // Only mark the beginnings of non-empty rows.
+        // Otherwise multiple rows could try to update the same key.
+        if (rowmap(i + 1) != rowBegin) {
+          keys(rowBegin) = uint64_t(i);
+        }
+      });
+  Kokkos::fence();
+  Kokkos::parallel_scan("CRS bulk sorting: compute keys", Kokkos::RangePolicy<ExecSpace>(exec, 0, entries.extent(0)),
+                        MaxScanFunctor<Offset, decltype(keys), Entries>(ncols, keys, entries));
+  Kokkos::fence();
+  return keys;
+}
+
+#ifndef KK_DISABLE_BULK_SORT_BY_KEY
+template <typename ExecSpace, typename Rowmap, typename Entries>
+Kokkos::View<typename Rowmap::non_const_value_type*, ExecSpace> computeEntryPermutation(
+    const ExecSpace& exec, const Rowmap& rowmap, const Entries& entries, typename Entries::non_const_value_type ncols) {
+  using Offset  = typename Rowmap::non_const_value_type;
+  using Ordinal = typename Entries::non_const_value_type;
```

Comment:
```suggestion
```

Reply:
Thanks, fixed now. For some reason nvcc+gcc with ``-Wunused-local-typedefs`` doesn't catch this, but a serial build with gcc does.